src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -971,6 +971,62 @@ CUDNN_SAFE_CALL(cudnnSetOpTensorDescriptor(opTensorDesc,
                 writer.block_end();
             }
 
+            template <>
+            void GPU_Emitter::EMITTER_DECL(ngraph::op::Reverse)
+            {
+                if (out[0].get_size() == 0)
+                {
+                    return;
+                }
+                auto reverse = static_cast<const op::Reverse*>(node);
+
+                const auto arg_shape = args[0].get_shape();
+                const auto arg_rank = arg_shape.size();
+                const auto result_shape = out[0].get_shape();
+                const auto reverse_axes = reverse->get_reversed_axes();
+                std::vector<size_t> reverse_axes_flag(arg_rank, 0);
+                for (auto a : reverse_axes)
+                {
+                    reverse_axes_flag[a] = 1;
+                }
+                writer.block_begin("  // " + node->get_name());
+                if (out[0].get_size() == 1)
+                {
+                    kernel::emit_memcpyDtD(writer, out[0], args[0]);
+                }
+                else
+                {
+                    writer << "size_t rank = " << arg_rank << ";\n";
+                    writer << "std::vector<size_t> input_shapes_h = {" << join(arg_shape, "UL,")
+                           << "UL};\n";
+                    writer << "std::vector<size_t> reverse_axes_h = {"
+                           << join(reverse_axes_flag, "UL,") << "UL};\n";
+
```

Comment:
Why not use the new CUDAEmitter instead of the old codegen API? Technically the old API is unsafe as it makes calls directly into nGraph from clang code. 

Additionally, I feel as though the codegen is getting in the way, because, as seen below, it forces us to do expensive cuda memcpys into the device from the JIT code rather than just passing them as arguments to the kernel from nGraph. I was planning to refactor the code that uses this old interface eventually, but am hoping that we can stop using the old API for new ops. Is that OK?

To some extent we can refactor in the second pass when we focus on performance, but it also doesn't make sense for us to have to redo those ops which haven't yet been implemented.

Reply:
Since this is working and already done, I'm not saying op::Reverse should be redone right now, just that if we can I personally think it would be better to use the new API moving forward.

Reply:
I see what you mean , good point.  Let's do it later.